devices/src/virtio/balloon.rs

```diff
@@ -1,245 +1,363 @@
-};
-use vm_control::{BalloonControlCommand, BalloonControlResponseSocket};
-use super::{
-    DescriptorChain, Queue, VirtioDevice, INTERRUPT_STATUS_CONFIG_CHANGED,
-    INTERRUPT_STATUS_USED_RING, TYPE_BALLOON, VIRTIO_F_VERSION_1,
-};
-#[derive(Debug)]
-pub enum BalloonError {
-    /// Request to adjust memory size can't provide the number of pages requested.
-    NotEnoughPages,
-    /// Failure wriitng the config notification event.
-    WritingConfigEvent(sys_util::Error),
-}
-pub type Result<T> = std::result::Result<T, BalloonError>;
-impl Display for BalloonError {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        use self::BalloonError::*;
-        match self {
-            NotEnoughPages => write!(f, "not enough pages"),
-            WritingConfigEvent(e) => write!(f, "failed to write config event: {}", e),
-        }
-    }
-}
+use sys_util::EventFd;
+use {DeviceEventT, EpollHandler};
+
 // Balloon has three virt IO queues: Inflate, Deflate, and Stats.
 // Stats is currently not used.
 const QUEUE_SIZE: u16 = 128;
 const QUEUE_SIZES: &[u16] = &[QUEUE_SIZE, QUEUE_SIZE];
 const VIRTIO_BALLOON_PFN_SHIFT: u32 = 12;
+
+const INTERRUPT_STATUS_USED_RING: u32 = 0x1;
+const INTERRUPT_STATUS_CONFIG_CHANGED: u32 = 0x2;
+
 // The feature bitmap for virtio balloon
 const VIRTIO_BALLOON_F_MUST_TELL_HOST: u32 = 0; // Tell before reclaiming pages
 const VIRTIO_BALLOON_F_DEFLATE_ON_OOM: u32 = 2; // Deflate balloon on OOM
-// BalloonConfig is modified by the worker and read from the device thread.
-#[derive(Default)]
-struct BalloonConfig {
-    num_pages: AtomicUsize,
-    actual_pages: AtomicUsize,
-}
-struct Worker {
+const VIRTIO_F_VERSION_1: u32 = 32;
+
+const BALLOON_INFLATE_EVENT: DeviceEventT = 0;
+const BALLOON_DEFLATE_EVENT: DeviceEventT = 1;
+pub const BALLOON_EVENTS_COUNT: usize = 2;
+
+/// Handler that drives the execution of the Balloon device
+pub struct BalloonEpollHandler {
     mem: GuestMemory,
     inflate_queue: Queue,
     deflate_queue: Queue,
     interrupt_status: Arc<AtomicUsize>,
+    inflate_queue_evt: EventFd,
+    deflate_queue_evt: EventFd,
     interrupt_evt: EventFd,
-    interrupt_resample_evt: EventFd,
-    config: Arc<BalloonConfig>,
-    command_socket: BalloonControlResponseSocket,
+    num_pages: Arc<AtomicUsize>,
 }
-fn valid_inflate_desc(desc: &DescriptorChain) -> bool {
-    !desc.is_write_only() && desc.len % 4 == 0
-}
-impl Worker {
-    fn process_inflate_deflate(&mut self, inflate: bool) -> bool {
-        let queue = if inflate {
-            &mut self.inflate_queue
+
+/// This takes a vector of page indices, and compacts them
+/// into ranges of consecutive pages. The result is a vector
+/// of (start_address, range_length) pairs.
+fn compact_page_indices(v: &mut Vec<u32>) -> Vec<(u32, u32)> {
+    // Simple special case.
+    if v.is_empty() {
+        return vec![];
+    }
+
+    // First sort the vector.
```

Comment:
I think it's worth mentioning that the number of indices to sort is limited and sorting this vec doesn't hurt complexity wise.